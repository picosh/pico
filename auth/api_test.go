package auth

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/picosh/pico/db"
	"github.com/picosh/pico/db/stub"
	"github.com/picosh/pico/shared"
)

var testUserID = "user-1"
var testUsername = "user-a"

func TestPaymentWebhook(t *testing.T) {
	apiConfig := setupTest()

	event := OrderEvent{
		Meta: &OrderEventMeta{
			EventName: "order_created",
			CustomData: &CustomDataMeta{
				PicoUsername: testUsername,
			},
		},
		Data: &OrderEventData{
			Attr: &OrderEventDataAttr{
				UserEmail:   "auth@pico.test",
				CreatedAt:   time.Now(),
				Status:      "paid",
				OrderNumber: 1337,
			},
		},
	}
	jso, err := json.Marshal(event)
	bail(err)
	hash := shared.HmacString(apiConfig.Cfg.SecretWebhook, string(jso))
	body := bytes.NewReader(jso)

	request := httptest.NewRequest("POST", mkpath("/webhook"), body)
	request.Header.Add("X-signature", hash)
	responseRecorder := httptest.NewRecorder()

	mux := authMux(apiConfig)
	mux.ServeHTTP(responseRecorder, request)

	testResponse(t, responseRecorder, 200, "text/plain")
}

func TestUser(t *testing.T) {
	apiConfig := setupTest()

	data := sishData{
		Username: testUsername,
	}
	jso, err := json.Marshal(data)
	bail(err)
	body := bytes.NewReader(jso)

	request := httptest.NewRequest("POST", mkpath("/user"), body)
	request.Header.Add("Authorization", "Bearer 123")
	responseRecorder := httptest.NewRecorder()

	mux := authMux(apiConfig)
	mux.ServeHTTP(responseRecorder, request)

	testResponse(t, responseRecorder, 200, "application/json")
}

func TestKey(t *testing.T) {
	apiConfig := setupTest()

	data := sishData{
		Username:  testUsername,
		PublicKey: "zzz",
	}
	jso, err := json.Marshal(data)
	bail(err)
	body := bytes.NewReader(jso)

	request := httptest.NewRequest("POST", mkpath("/key"), body)
	request.Header.Add("Authorization", "Bearer 123")
	responseRecorder := httptest.NewRecorder()

	mux := authMux(apiConfig)
	mux.ServeHTTP(responseRecorder, request)

	testResponse(t, responseRecorder, 200, "application/json")
}

func TestCheckout(t *testing.T) {
	apiConfig := setupTest()

	request := httptest.NewRequest("GET", mkpath("/checkout/"+testUsername), strings.NewReader(""))
	request.Header.Add("Authorization", "Bearer 123")
	responseRecorder := httptest.NewRecorder()

	mux := authMux(apiConfig)
	mux.ServeHTTP(responseRecorder, request)

	loc := responseRecorder.Header().Get("Location")
	if loc != "https://checkout.pico.sh/buy/73c26cf9-3fac-44c3-b744-298b3032a96b?discount=0&checkout[custom][username]=user-a" {
		t.Errorf("Have Location %s, want checkout", loc)
	}
	if responseRecorder.Code != http.StatusMovedPermanently {
		t.Errorf("Want status '%d', got '%d'", http.StatusMovedPermanently, responseRecorder.Code)
		return
	}
}

func TestIntrospect(t *testing.T) {
	apiConfig := setupTest()

	request := httptest.NewRequest("POST", mkpath("/introspect?token=123"), strings.NewReader(""))
	responseRecorder := httptest.NewRecorder()

	mux := authMux(apiConfig)
	mux.ServeHTTP(responseRecorder, request)

	testResponse(t, responseRecorder, 200, "application/json")
}

func TestToken(t *testing.T) {
	apiConfig := setupTest()

	request := httptest.NewRequest("POST", mkpath("/token?code=123"), strings.NewReader(""))
	responseRecorder := httptest.NewRecorder()

	mux := authMux(apiConfig)
	mux.ServeHTTP(responseRecorder, request)

	testResponse(t, responseRecorder, 200, "application/json")
}

func TestAuthApi(t *testing.T) {
	apiConfig := setupTest()
	tt := []*ApiExample{
		{
			name:        "authorize",
			path:        "/authorize?response_type=json&client_id=333&redirect_uri=pico.test&scope=admin",
			status:      http.StatusOK,
			contentType: "text/html; charset=utf-8",
			dbpool:      apiConfig.Dbpool,
		},
		{
			name:        "rss",
			path:        "/rss/123",
			status:      http.StatusOK,
			contentType: "application/atom+xml",
			dbpool:      apiConfig.Dbpool,
		},
		{
			name:        "fileserver",
			path:        "/robots.txt",
			status:      http.StatusOK,
			contentType: "text/plain; charset=utf-8",
			dbpool:      apiConfig.Dbpool,
		},
		{
			name:        "well-known",
			path:        "/.well-known/oauth-authorization-server",
			status:      http.StatusOK,
			contentType: "application/json",
			dbpool:      apiConfig.Dbpool,
		},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			request := httptest.NewRequest("GET", mkpath(tc.path), strings.NewReader(""))
			responseRecorder := httptest.NewRecorder()

			mux := authMux(apiConfig)
			mux.ServeHTTP(responseRecorder, request)

			testResponse(t, responseRecorder, tc.status, tc.contentType)
		})
	}
}

type ApiExample struct {
	name        string
	path        string
	status      int
	contentType string
	dbpool      db.DB
}

type AuthDb struct {
	*stub.StubDB
}

func (a *AuthDb) AddPicoPlusUser(username, from, txid string) error {
	return nil
}

func (a *AuthDb) FindUserForName(username string) (*db.User, error) {
	return &db.User{ID: testUserID, Name: username}, nil
}

func (a *AuthDb) FindUserForKey(username string, pubkey string) (*db.User, error) {
	return &db.User{ID: testUserID, Name: username}, nil
}

func (a *AuthDb) FindUserForToken(token string) (*db.User, error) {
	if token != "123" {
		return nil, fmt.Errorf("invalid token")
	}
	return &db.User{ID: testUserID, Name: testUsername}, nil
}

func (a *AuthDb) HasFeatureForUser(userID string, feature string) bool {
	return true
}

func (a *AuthDb) FindKeysForUser(user *db.User) ([]*db.PublicKey, error) {
	return []*db.PublicKey{{ID: "1", UserID: user.ID, Name: "my-key", Key: "nice-pubkey", CreatedAt: &time.Time{}}}, nil
}

func (a *AuthDb) FindFeatureForUser(userID string, feature string) (*db.FeatureFlag, error) {
	now := time.Date(2021, 8, 15, 14, 30, 45, 100, time.UTC)
	oneDayWarning := now.AddDate(0, 0, 1)
	return &db.FeatureFlag{ID: "2", UserID: userID, Name: "plus", ExpiresAt: &oneDayWarning, CreatedAt: &now}, nil
}

func NewAuthDb(logger *slog.Logger) *AuthDb {
	sb := stub.NewStubDB(logger)
	return &AuthDb{
		StubDB: sb,
	}
}

func mkpath(path string) string {
	return fmt.Sprintf("https://auth.pico.test%s", path)
}

func setupTest() *shared.ApiConfig {
	logger := shared.CreateLogger("auth")
	cfg := &shared.ConfigSite{
		Issuer:        "auth.pico.test",
		Domain:        "http://0.0.0.0:3000",
		Port:          "3000",
		Secret:        "",
		SecretWebhook: "my-secret",
	}
	cfg.Logger = logger
	db := NewAuthDb(cfg.Logger)
	apiConfig := &shared.ApiConfig{
		Cfg:    cfg,
		Dbpool: db,
	}

	return apiConfig
}

func testResponse(t *testing.T, responseRecorder *httptest.ResponseRecorder, status int, contentType string) {
	if responseRecorder.Code != status {
		t.Errorf("Want status '%d', got '%d'", status, responseRecorder.Code)
		return
	}

	ct := responseRecorder.Header().Get("content-type")
	if ct != contentType {
		t.Errorf("Want content type '%s', got '%s'", contentType, ct)
		return
	}

	body := strings.TrimSpace(responseRecorder.Body.String())
	snaps.MatchSnapshot(t, body)
}

func bail(err error) {
	if err != nil {
		panic(bail)
	}
}
