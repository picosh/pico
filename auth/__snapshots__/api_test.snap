
[TestPaymentWebhook - 1]
successfully added pico+ user
---

[TestAuthApi/authorize - 1]
<!doctype html>
<html lang="en">
  <head>
    <meta charset='utf-8'>
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>auth redirect</title>

    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">

    <meta name="keywords" content="static, site, hosting" />

    <link rel="stylesheet" href="/main.css" />

    
  </head>
  <body >
<header>
    <h1 class="text-2xl">Auth Redirect</h1>
    <hr />
</header>
<main>
    <section>
        <h2 class="text-xl">You are being redirected to pico.test</h2>
        <p>
            Here is their auth request data:
        </p>

        <article>
            <h2 class="text-lg">Client ID</h2>
            <div>333</div>
        </article>

        <br />

        <article>
            <h2 class="text-lg">Redirect URI</h2>
            <div>pico.test</div>
        </article>

        <br />

        <article>
            <h2 class="text-lg">Scope</h2>
            <div>admin</div>
        </article>

        <br />

        <article>
            <h2 class="text-lg">Response Type</h2>
            <div>json</div>
        </article>

        <br />

        <article>
            <h2 class="text-lg">If you would like to continue authenticating with this service, ssh into a pico service and generate a token. Then input it here and click submit.</h2>
            <br />
            <form action="/redirect" method="POST">
                <label for="token">Auth Token:</label><br>
                <input type="text" id="token" name="token"><br>
                <br />
                <input type="hidden" id="redirect_uri" name="redirect_uri" value="pico.test">
                <input type="hidden" id="response_type" name="response_type" value="json">
                <input type="submit" value="Submit">
            </form>
        </article>
    </section>
</main>

<footer>
  <hr />
  <p class="font-italic">Built and maintained by <a href="https://pico.sh">pico.sh</a>.</p>
  <div>
    <a href="https://github.com/picosh/pico">source</a>
  </div>
</footer>

</body>
</html>
---

[TestUser - 1]
[{"id":"1","user_id":"user-1","name":"my-key","key":"nice-pubkey","created_at":"0001-01-01T00:00:00Z"}]
---

[TestAuthApi/rss - 1]
<?xml version="1.0" encoding="UTF-8"?><feed xmlns="http://www.w3.org/2005/Atom">
  <title>pico+</title>
  <id>https://auth.pico.sh/rss/123</id>
  <updated></updated>
  <subtitle>get notified of important membership updates</subtitle>
  <link href="https://auth.pico.sh/rss/123"></link>
  <author>
    <name>team pico</name>
  </author>
  <entry>
    <title>pico+ membership activated on 2021-08-15 14:30:45</title>
    <updated>2021-08-15T14:30:45Z</updated>
    <id>pico-plus-activated-1629037845</id>
    <content type="html">Thanks for joining pico+! You now have access to all our premium services for exactly one year.  We will send you pico+ expiration notifications through this RSS feed.  Go to &lt;a href=&#34;https://pico.sh/getting-started#next-steps&#34;&gt;pico.sh/getting-started#next-steps&lt;/a&gt; to start using our services.</content>
    <link href="https://pico.sh" rel="alternate"></link>
    <summary type="html">Thanks for joining pico+! You now have access to all our premium services for exactly one year.  We will send you pico+ expiration notifications through this RSS feed.  Go to &lt;a href=&#34;https://pico.sh/getting-started#next-steps&#34;&gt;pico.sh/getting-started#next-steps&lt;/a&gt; to start using our services.</summary>
    <author>
      <name>team pico</name>
    </author>
  </entry>
  <entry>
    <title>pico+ 1-month expiration notice</title>
    <updated>2021-07-16T14:30:45Z</updated>
    <id>1626445845</id>
    <content type="html">Your pico+ membership is going to expire on 2021-08-16 14:30:45</content>
    <link href="https://pico.sh" rel="alternate"></link>
    <summary type="html">Your pico+ membership is going to expire on 2021-08-16 14:30:45</summary>
    <author>
      <name>team pico</name>
    </author>
  </entry>
  <entry>
    <title>pico+ 1-week expiration notice</title>
    <updated>2021-08-09T14:30:45Z</updated>
    <id>1628519445</id>
    <content type="html">Your pico+ membership is going to expire on 2021-08-16 14:30:45</content>
    <link href="https://pico.sh" rel="alternate"></link>
    <summary type="html">Your pico+ membership is going to expire on 2021-08-16 14:30:45</summary>
    <author>
      <name>team pico</name>
    </author>
  </entry>
  <entry>
    <title>pico+ 1-day expiration notice</title>
    <updated>2021-08-14T14:30:45Z</updated>
    <id>1628951445</id>
    <content type="html">Your pico+ membership is going to expire on 2021-08-16 14:30:45</content>
    <link href="https://pico.sh" rel="alternate"></link>
    <summary type="html">Your pico+ membership is going to expire on 2021-08-16 14:30:45</summary>
    <author>
      <name>team pico</name>
    </author>
  </entry>
</feed>
---

[TestKey - 1]
{"id":"user-1","name":"user-a","created_at":null}
---

[TestAuthApi/fileserver - 1]
User-agent: *
Allow: /
---

[TestAuthApi/well-known - 1]
{"issuer":"auth.pico.test","introspection_endpoint":"http://0.0.0.0:3000/introspect","introspection_endpoint_auth_methods_supported":["none"],"authorization_endpoint":"http://0.0.0.0:3000/authorize","token_endpoint":"http://0.0.0.0:3000/token","response_types_supported":["code"]}
---

[TestIntrospect - 1]
{"active":true,"username":"user-a"}
---

[TestToken - 1]
{"access_token":"123"}
---
